privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+            pubkey = _strip_key(self.get_tokeninfo(PUBLIC_KEY_SERVER))
+            response_detail["public_key"] = pubkey
+
+        return response_detail
+
+    @classmethod
+    def api_endpoint(cls, request, g):
+        """
+        This provides a function which is called by the API endpoint
+        /ttype/push which is defined in api/ttype.py
+
+        The method returns
+            return "json", {}
+
+        This endpoint is used for the 2nd enrollment step of the smartphone.
+        Parameters sent:
+            * serial
+            * fbtoken
+            * pubkey
+
+        This endpoint is also used, if the smartphone sends the signed response
+        to the challenge during authentication
+        Parameters sent:
+            * serial
+            * nonce (which is the challenge)
+            * signature (which is the signed nonce)
+
+
+        :param request: The Flask request
+        :param g: The Flask global object g
+        :return: dictionary
+        """
+        from privacyidea.lib.token import get_tokens
+        from privacyidea.lib.utils import prepare_result
+        details = {}
+        result = False
+        serial = getParam(request.all_data, "serial", optional=False)
+
+        if serial and "fbtoken" in request.all_data and "pubkey" in request.all_data:
+            # Do the 2nd step of the enrollment
+            toks = get_tokens(serial=serial, rollout_state="clientwait")
+            if len(toks) == 0:
+                raise ParameterError("No token with this serial number in the rollout state 'clientwait'.")
+            token_obj = toks[0]
+
+            token_obj.update(request.all_data)
+            init_detail_dict = request.all_data
+
+            details = token_obj.get_init_detail(init_detail_dict)
+            result = True
+        elif serial and "nonce" in request.all_data and "signature" in request.all_data:
+            challenge = getParam(request.all_data, "nonce")
+            serial = getParam(request.all_data, "serial")
+            signature = getParam(request.all_data, "signature")
+
+            # get the token_obj for the given serial:
+            toks = get_tokens(serial=serial)
+            token_obj = toks[0]
+            pubkey_pem = token_obj.get_tokeninfo(PUBLIC_KEY_SMARTPHONE)
+            if not pubkey_pem.startswith("-----"):
+                pubkey_pem = "-----BEGIN PUBLIC KEY-----\n{0!s}\n-----END PUBLIC KEY-----".format(pubkey_pem.strip().replace(" ", "+"))
+            # Do the 2nd step of the authentication
+            # Find valid challenges
+            challengeobject_list = get_challenges(serial=serial, challenge=challenge)
+
+            if challengeobject_list:
+                # There are valid challenges, so we check this signature
+                for chal in challengeobject_list:
+                    # verify the signature of the nonce
+                    pubkey_obj = serialization.load_pem_public_key(str(pubkey_pem), default_backend())
```

Comment:
For Python 3 compatibility, we need to use ``to_bytes(pubkey_pem)`` here instead of ``str(...)`` -- the latter will produce a (unicode) string, which is not accepted by ``load_pem_public_key``.

Reply:
done in c3f36915